File: android/Gutenberg/src/main/java/org/wordpress/gutenberg/http/HTTPRequestSerializer.kt
package org.wordpress.gutenberg.http

/**
 * Errors thrown when parsing an HTTP/1.1 request fails due to RFC 7230/9112 violations.
 */
enum class HTTPRequestParseError(
    /** The HTTP status code that should be sent for this error. */
    val httpStatus: Int,
    /** A camelCase identifier matching the Swift error case names and JSON fixture keys. */
    val errorId: String
) {
    EMPTY_HEADER_SECTION(400, "emptyHeaderSection"),
    MALFORMED_REQUEST_LINE(400, "malformedRequestLine"),
    OBS_FOLD_DETECTED(400, "obsFoldDetected"),
    WHITESPACE_BEFORE_COLON(400, "whitespaceBeforeColon"),
    INVALID_CONTENT_LENGTH(400, "invalidContentLength"),
    CONFLICTING_CONTENT_LENGTH(400, "conflictingContentLength"),
    UNSUPPORTED_TRANSFER_ENCODING(400, "unsupportedTransferEncoding"),
    INVALID_HTTP_VERSION(400, "invalidHTTPVersion"),
    INVALID_FIELD_NAME(400, "invalidFieldName"),
    INVALID_FIELD_VALUE(400, "invalidFieldValue"),
    MISSING_HOST_HEADER(400, "missingHostHeader"),
    MULTIPLE_HOST_HEADERS(400, "multipleHostHeaders"),
    PAYLOAD_TOO_LARGE(413, "payloadTooLarge"),
    HEADERS_TOO_LARGE(431, "headersTooLarge"),
    TOO_MANY_HEADERS(431, "tooManyHeaders"),
    INVALID_ENCODING(400, "invalidEncoding"),
    BUFFER_IO_ERROR(500, "bufferIOError");
}

/**
 * Exception thrown when HTTP request parsing fails.
 */
class HTTPRequestParseException(val error: HTTPRequestParseError) : Exception(error.errorId)

/**
 * Parsed header information from an HTTP request.
 */
data class ParsedHeaders(
    /** The HTTP method (e.g., "GET", "POST"). */
    val method: String,
    /** The request-target from the HTTP request line, per RFC 9112 Section 3. */
    val target: String,
    /** The HTTP-version from the request line (e.g., "HTTP/1.1"), per RFC 9112 §2.3. */
    val httpVersion: String,
    /** The HTTP headers as key-value pairs, preserving original casing. */
    val headers: LinkedHashMap<String, String>,
    /** The value of the `Content-Length` header, or 0 if absent. */
    val contentLength: Long,
    /** The byte offset where the body begins (after the `\r\n\r\n` separator). */
    val bodyOffset: Int
)

/**
 * The result of attempting to parse HTTP request headers.
 */
sealed class HeaderParseResult {
    /** The data does not yet contain the complete header section (`\r\n\r\n`). */
    data object NeedsMoreData : HeaderParseResult()
    /** The header data is malformed and cannot be parsed. */
    data class Invalid(val error: HTTPRequestParseError) : HeaderParseResult()
    /** Headers were successfully parsed. */
    data class Parsed(val headers: ParsedHeaders) : HeaderParseResult()
}

/**
 * Parses raw HTTP/1.1 request bytes into structured components.
 *
 * This is the stateless parser that converts raw HTTP request data into
 * a [ParsedHeaders] structure. For incremental parsing with buffering,
 * use [HTTPRequestParser] instead.
 */
internal object HTTPRequestSerializer {

    /**
     * Attempts to parse the HTTP request line and headers from raw data.
     *
     * Looks for the `\r\n\r\n` header terminator, then parses the request line
     * and individual headers. Returns [HeaderParseResult.NeedsMoreData] if the
     * terminator hasn't been received yet, or [HeaderParseResult.Invalid] with
     * a specific error if the request is malformed.
     */
    fun parseHeaders(data: ByteArray): HeaderParseResult {
        // RFC 7230 §3.5: Skip leading CRLFs for robustness.
        var scanOffset = 0
        while (scanOffset + 1 < data.size &&
            data[scanOffset] == 0x0D.toByte() &&
            data[scanOffset + 1] == 0x0A.toByte()
        ) {
            scanOffset += 2
        }
        if (scanOffset >= data.size) return HeaderParseResult.NeedsMoreData

        val separator = "\r\n\r\n".toByteArray(Charsets.UTF_8)
        val separatorIndex = ReadOnlyBytes(data).indexOf(separator, scanOffset)
        if (separatorIndex == -1) return HeaderParseResult.NeedsMoreData

        val headerBytes = data.copyOfRange(scanOffset, separatorIndex)

        // Validate UTF-8 encoding by round-tripping through String and back.
        val headerString: String
        try {
            headerString = headerBytes.toString(Charsets.UTF_8)
            val reEncoded = headerString.toByteArray(Charsets.UTF_8)
            if (!reEncoded.contentEquals(headerBytes)) {
                return HeaderParseResult.Invalid(HTTPRequestParseError.INVALID_ENCODING)
            }
        } catch (_: Exception) {
            return HeaderParseResult.Invalid(HTTPRequestParseError.INVALID_ENCODING)
        }

        val lines = headerString.split("\r\n")
        val requestLine = lines.firstOrNull()
        if (requestLine.isNullOrEmpty()) {
            return HeaderParseResult.Invalid(HTTPRequestParseError.EMPTY_HEADER_SECTION)
        }

        val parts = requestLine.split(" ", limit = 3)
        if (parts.size < 2) {
            return HeaderParseResult.Invalid(HTTPRequestParseError.MALFORMED_REQUEST_LINE)
        }

        val method = parts[0]
        val target = parts[1]

        // RFC 9110 §9.1: method = token (tchar characters only).
        if (!method.all { isTokenChar(it) }) {
            return HeaderParseResult.Invalid(HTTPRequestParseError.MALFORMED_REQUEST_LINE)
        }

        // RFC 9112 §2.3: HTTP-version = "HTTP/" DIGIT "." DIGIT
        if (parts.size < 3) {
            return HeaderParseResult.Invalid(HTTPRequestParseError.INVALID_HTTP_VERSION)
        }
        val httpVersion = parts[2]
        if (!isValidHTTPVersion(httpVersion)) {
            return HeaderParseResult.Invalid(HTTPRequestParseError.INVALID_HTTP_VERSION)
        }

        // RFC 9112 §3.2: Validate request-target form.
        if (method == "CONNECT") {
            if (target.startsWith("/") || !target.contains(":")) {
                return HeaderParseResult.Invalid(HTTPRequestParseError.MALFORMED_REQUEST_LINE)
            }
        } else if (method == "OPTIONS" && target == "*") {
            // asterisk-form is valid for OPTIONS
        } else if (target.startsWith("/")) {
            // origin-form — valid for all methods
        } else if (target.lowercase().startsWith("http://") || target.lowercase().startsWith("https://")) {
            // absolute-form — valid for all methods
        } else {
            return HeaderParseResult.Invalid(HTTPRequestParseError.MALFORMED_REQUEST_LINE)
        }

        val headers = LinkedHashMap<String, String>()
        val keyIndex = HashMap<String, String>()  // lowercased -> original casing
        var contentLengthValue: Long? = null
        var hostHeaderCount = 0
        var headerCount = 0

        for (line in lines.drop(1)) {
            if (line.isEmpty()) continue

            headerCount++
            if (headerCount > 100) {
                return HeaderParseResult.Invalid(HTTPRequestParseError.TOO_MANY_HEADERS)
            }

            // RFC 7230 §3.2.4: Reject obs-fold
            if (line[0] == ' ' || line[0] == '\t') {
                return HeaderParseResult.Invalid(HTTPRequestParseError.OBS_FOLD_DETECTED)
            }

            val colonIndex = line.indexOf(':')
            if (colonIndex == -1) {
                return HeaderParseResult.Invalid(HTTPRequestParseError.INVALID_FIELD_NAME)
            }

            val rawKey = line.substring(0, colonIndex)

            // RFC 7230 §3.2.4: No whitespace is allowed between the field-name and colon.
            // Check this before the general token validation so we return the more
            // specific error (WHITESPACE_BEFORE_COLON) instead of INVALID_FIELD_NAME.
            if (rawKey.any { it == ' ' || it == '\t' }) {
                return HeaderParseResult.Invalid(HTTPRequestParseError.WHITESPACE_BEFORE_COLON)
            }

            // RFC 9110 §5.1: field-name = token
            if (rawKey.isEmpty() || !rawKey.all { isTokenChar(it) }) {
                return HeaderParseResult.Invalid(HTTPRequestParseError.INVALID_FIELD_NAME)
            }

            val key = rawKey
            val lowerKey = key.lowercase()
            // RFC 9110 §5.5: OWS (optional whitespace) is SP / HTAB only.
            // Kotlin's String.trim() strips all chars <= 0x20 (including CR, LF, etc.),
            // which would silently remove bare CRs before field value validation.
            val value = line.substring(colonIndex + 1).trimOWS()

            // RFC 9110 §5.5: Validate field value characters.
            for (c in value) {
                val v = c.code
                if (v <= 0x08 || (v in 0x0A..0x1F) || v == 0x7F) {
                    return HeaderParseResult.Invalid(HTTPRequestParseError.INVALID_FIELD_VALUE)
                }
            }

            // RFC 7230 §3.3.3: Reject Transfer-Encoding
            if (lowerKey == "transfer-encoding") {
                return HeaderParseResult.Invalid(HTTPRequestParseError.UNSUPPORTED_TRANSFER_ENCODING)
            }

            // Content-Length: validate and normalize
            if (lowerKey == "content-length") {
                contentLengthValue = try {
                    validateContentLength(value, contentLengthValue)
                } catch (e: HTTPRequestParseException) {
                    return HeaderParseResult.Invalid(e.error)
                }
                val resolved = contentLengthValue.toString()
                val existingKey = keyIndex["content-length"]
                if (existingKey != null) {
                    headers[existingKey] = resolved
                } else {
                    headers[key] = resolved
                    keyIndex["content-length"] = key
                }
                continue
            }

            // Track Host header occurrences
            if (lowerKey == "host") {
                hostHeaderCount++
            }

            // RFC 9110 §5.3: Combine duplicate field lines with comma-separated values.
            val existingKey = keyIndex[lowerKey]
            if (existingKey != null) {
                headers[existingKey] = "${headers[existingKey]}, $value"
            } else {
                headers[key] = value
                keyIndex[lowerKey] = key
            }
        }

        // RFC 9110 §7.2: Host header validation
        if (hostHeaderCount > 1) {
            return HeaderParseResult.Invalid(HTTPRequestParseError.MULTIPLE_HOST_HEADERS)
        }
        if (httpVersion == "HTTP/1.1" && hostHeaderCount == 0) {
            return HeaderParseResult.Invalid(HTTPRequestParseError.MISSING_HOST_HEADER)
        }

        val contentLength = contentLengthValue ?: 0L
        val bodyOffset = separatorIndex + separator.size

        return HeaderParseResult.Parsed(
            ParsedHeaders(
                method = method,
                target = target,
                httpVersion = httpVersion,
                headers = headers,
                contentLength = contentLength,
                bodyOffset = bodyOffset
            )
        )
    }

    /**
     * Validates a Content-Length header value per RFC 9110 §8.6 / RFC 7230 §3.3.3.
     */
    internal fun validateContentLength(value: String, existing: Long?): Long {
        val parts = value.split(",").map { it.trim() }
        val first = parts.firstOrNull()
        if (first.isNullOrEmpty() || !first.all { isAsciiDigit(it) }) {
            throw HTTPRequestParseException(HTTPRequestParseError.INVALID_CONTENT_LENGTH)
        }
        val cl = first.toLongOrNull()
            ?: throw HTTPRequestParseException(HTTPRequestParseError.INVALID_CONTENT_LENGTH)
        if (cl < 0) throw HTTPRequestParseException(HTTPRequestParseError.INVALID_CONTENT_LENGTH)

        for (part in parts.drop(1)) {
            if (part.isEmpty() || !part.all { isAsciiDigit(it) }) {
                throw HTTPRequestParseException(HTTPRequestParseError.CONFLICTING_CONTENT_LENGTH)
            }
            val partValue = part.toLongOrNull()
            if (partValue != cl) {
                throw HTTPRequestParseException(HTTPRequestParseError.CONFLICTING_CONTENT_LENGTH)
            }
        }

        if (existing != null && existing != cl) {
            throw HTTPRequestParseException(HTTPRequestParseError.CONFLICTING_CONTENT_LENGTH)
        }
        return cl
    }

    /**
     * Validates that a string matches the HTTP-version format: `HTTP/DIGIT.DIGIT`.
     */
    private fun isValidHTTPVersion(version: String): Boolean {
        if (!version.startsWith("HTTP/")) return false
        val rest = version.removePrefix("HTTP/")
        val parts = rest.split(".", limit = 2)
        if (parts.size != 2) return false
        return parts[0].length == 1 && parts[0][0].isDigit() &&
            parts[1].length == 1 && parts[1][0].isDigit()
    }

    /**
     * Returns whether a character is a valid HTTP token character (RFC 9110 §5.6.2).
     */
    private fun isTokenChar(c: Char): Boolean {
        val ascii = c.code
        return when {
            ascii in 0x41..0x5A -> true // A-Z
            ascii in 0x61..0x7A -> true // a-z
            ascii in 0x30..0x39 -> true // 0-9
            c in "!#\$%&'*+-.^_`|~" -> true
            else -> false
        }
    }

    /**
     * Returns whether a character is an ASCII digit (0-9).
     * Unlike [Char.isDigit], this rejects non-ASCII Unicode digits.
     */
    private fun isAsciiDigit(c: Char): Boolean = c.code in 0x30..0x39

}

/**
 * Trims only SP (0x20) and HTAB (0x09) from both ends of a string.
 *
 * This matches RFC 9110's OWS (optional whitespace) definition:
 *   OWS = *( SP / HTAB )
 *
 * Unlike Kotlin's [String.trim], this does NOT strip CR, LF, or other
 * control characters, ensuring they are preserved for field value
 * validation (which must reject them per RFC 9110 §5.5).
 */
internal fun String.trimOWS(): String {
    var start = 0
    var end = length
    while (start < end && (this[start] == ' ' || this[start] == '\t')) start++
    while (end > start && (this[end - 1] == ' ' || this[end - 1] == '\t')) end--
    return if (start == 0 && end == length) this else substring(start, end)
}

File: ios/Sources/GutenbergKitHTTP/HTTPRequestSerializer.swift
import Foundation

/// Errors thrown when parsing an HTTP/1.1 request fails due to RFC 7230/9112 violations.
public enum HTTPRequestParseError: Error, Sendable, Equatable, LocalizedError, CaseIterable {
    /// The header section before `\r\n\r\n` is empty (e.g., `\r\n\r\n` with no request line).
    case emptyHeaderSection
    /// The request line does not contain at least a method and target (RFC 9112 §3).
    case malformedRequestLine
    /// A header line starts with SP or HTAB, indicating obs-fold (RFC 7230 §3.2.4).
    case obsFoldDetected
    /// A header field-name contains whitespace before the colon (RFC 7230 §3.2.4).
    case whitespaceBeforeColon
    /// The `Content-Length` value is not a valid non-negative integer (RFC 9112 §6.3).
    case invalidContentLength
    /// Multiple `Content-Length` headers with conflicting values (RFC 7230 §3.3.3).
    case conflictingContentLength
    /// The request contains a `Transfer-Encoding` header, which is not supported (RFC 7230 §3.3.3).
    case unsupportedTransferEncoding
    /// The HTTP-version in the request line is missing or does not match `HTTP/DIGIT.DIGIT` (RFC 9112 §2.3).
    case invalidHTTPVersion
    /// A header field-name is empty or contains characters outside the token set (RFC 9110 §5.1).
    case invalidFieldName
    /// A header field value contains NUL (0x00) or bare CR (RFC 9110 §5.5).
    case invalidFieldValue
    /// An HTTP/1.1 request is missing the required Host header (RFC 9110 §7.2).
    case missingHostHeader
    /// The request contains more than one Host header field line (RFC 9110 §7.2).
    case multipleHostHeaders
    /// The `Content-Length` exceeds the server's maximum allowed body size.
    case payloadTooLarge
    /// The header section exceeds the maximum allowed size before the terminator is found.
    case headersTooLarge
    /// The request contains more than 100 header field lines.
    case tooManyHeaders
    /// The header data could not be decoded as UTF-8.
    case invalidEncoding
    /// An I/O error occurred while buffering the request (e.g. disk full).
    case bufferIOError

    /// How the parser disposes of a parse error.
    public enum Disposition: Sendable {
        /// Abort the connection; the malformed request never reaches the handler.
        case fatal
        /// Surface to the handler via ``HTTPRequestParser/parseError`` so it can
        /// build a response.
        case recoverable
    }

    /// Whether this error aborts the connection (``Disposition/fatal``) or is
    /// surfaced to the handler (``Disposition/recoverable``).
    ///
    /// Only genuinely recoverable errors — where the request line and headers are
    /// well-formed — may be recoverable; anything smuggling-relevant (framing,
    /// Content-Length) must stay fatal so the request never reaches the handler.
    public var disposition: Disposition {
        switch self {
        case .payloadTooLarge:
            return .recoverable
        case .emptyHeaderSection, .malformedRequestLine, .obsFoldDetected,
             .whitespaceBeforeColon, .invalidContentLength, .conflictingContentLength,
             .unsupportedTransferEncoding, .invalidHTTPVersion, .invalidFieldName,
             .invalidFieldValue, .missingHostHeader, .multipleHostHeaders,
             .headersTooLarge, .tooManyHeaders, .invalidEncoding, .bufferIOError:
            return .fatal
        }
    }

    /// The HTTP status code that should be sent for this error.
    public var httpStatus: Int {
        switch self {
        case .payloadTooLarge: return 413
        case .headersTooLarge, .tooManyHeaders: return 431
        case .bufferIOError: return 500
        default: return 400
        }
    }

    /// The HTTP reason phrase for the corresponding status code.
    ///
    /// Returns a `StaticString` so callers needing a stable C pointer (e.g., JNI bridge)
    /// can use `utf8Start` without allocation.
    public var httpStatusText: StaticString {
        HTTPResponse.defaultStatusText(for: httpStatus)
    }

    public var errorDescription: String? {
        switch self {
        case .emptyHeaderSection:
            return "The HTTP request header section is empty — no request line was found before the header terminator."
        case .malformedRequestLine:
            return "The HTTP request line is malformed — it must contain at least a method and request target separated by a space."
        case .obsFoldDetected:
            return "The HTTP request contains an obsolete line folding (a header continuation line starting with a space or tab), which is rejected per RFC 7230 §3.2.4."
        case .whitespaceBeforeColon:
            return "A header field name contains whitespace before the colon, which is rejected per RFC 7230 §3.2.4 to prevent request smuggling."
        case .invalidContentLength:
            return "The Content-Length header value is not a valid non-negative integer."
        case .conflictingContentLength:
            return "The request contains multiple Content-Length headers with different values, which is rejected per RFC 7230 §3.3.3."
        case .unsupportedTransferEncoding:
            return "The request contains a Transfer-Encoding header. This server only supports identity encoding with Content-Length framing (RFC 7230 §3.3.3)."
        case .invalidHTTPVersion:
            return "The HTTP-version in the request line is missing or malformed — it must match HTTP/DIGIT.DIGIT (e.g., HTTP/1.1) per RFC 9112 §2.3."
        case .invalidFieldName:
            return "A header field name is empty or contains characters outside the HTTP token set per RFC 9110 §5.1."
        case .invalidFieldValue:
            return "A header field value contains a NUL byte (0x00) or bare CR, which is rejected per RFC 9110 §5.5."
        case .missingHostHeader:
            return "The HTTP/1.1 request is missing the required Host header field per RFC 9110 §7.2."
        case .multipleHostHeaders:
            return "The request contains more than one Host header field line, which is rejected per RFC 9110 §7.2."
        case .payloadTooLarge:
            return "The request body exceeds the maximum allowed size."
        case .headersTooLarge:
            return "The request header section exceeds the maximum allowed size (64 KB)."
        case .tooManyHeaders:
            return "The request contains more than 100 header field lines."
        case .invalidEncoding:
            return "The HTTP request headers could not be decoded as UTF-8."
        case .bufferIOError:
            return "An I/O error occurred while buffering the request body (e.g. disk full)."
        }
    }
}

/// Parses raw HTTP/1.1 request bytes into structured components.
///
/// This is the stateless counterpart to ``HTTPResponseSerializer`` — it converts
/// raw HTTP request data into a ``ParsedHTTPRequest``, while ``HTTPResponseSerializer``
/// converts structured response data into raw bytes.
///
/// For incremental parsing with buffering, use ``HTTPRequestParser`` instead.
///
/// ```swift
/// let data = Data("GET /api HTTP/1.1\r\nHost: localhost\r\n\r\n".utf8)
/// if case .parsed(let headers) = HTTPRequestSerializer.parseHeaders(from: data) {
///     print(headers.method, headers.target)
/// }
/// ```
public enum HTTPRequestSerializer {

    /// Parsed header information from an HTTP request.
    public struct ParsedHeaders: Sendable {
        /// The HTTP method (e.g., "GET", "POST").
        public let method: String
        /// The request-target from the HTTP request line, per RFC 9112 Section 3.
        public let target: String
        /// The HTTP-version from the request line (e.g., "HTTP/1.1"), per RFC 9112 §2.3.
        public let httpVersion: String
        /// The HTTP headers as key-value pairs, preserving original casing.
        public let headers: [String: String]
        /// The value of the `Content-Length` header, or 0 if absent.
        public let contentLength: Int64
        /// The byte offset where the body begins (after the `\r\n\r\n` separator).
        public let bodyOffset: Int
    }

    /// The result of attempting to parse HTTP request headers.
    public enum HeaderParseResult: Sendable {
        /// The data does not yet contain the complete header section (`\r\n\r\n`).
        case needsMoreData
        /// The header data is malformed and cannot be parsed.
        case invalid(HTTPRequestParseError)
        /// Headers were successfully parsed.
        case parsed(ParsedHeaders)
    }

    /// Attempts to parse the HTTP request line and headers from raw data.
    ///
    /// Looks for the `\r\n\r\n` header terminator, then parses the request line
    /// and individual headers. Returns `.needsMoreData` if the terminator hasn't
    /// been received yet, or `.invalid` with a specific ``HTTPRequestParseError``
    /// if the request is malformed.
    ///
    /// - Parameter data: The raw HTTP request bytes received so far.
    /// - Returns: The parse result indicating whether headers are available.
    public static func parseHeaders(from data: Data) -> HeaderParseResult {
        // Ensure zero-based indexing — Data slices retain their original indices,
        // so a caller passing e.g. `fullData[500...]` would crash on `data[0]`.
        let data = Data(data)

        // RFC 7230 §3.5: Skip leading CRLFs for robustness.
        // A server SHOULD ignore at least one empty line received prior to the request-line.
        var scanOffset = 0
        while scanOffset + 1 < data.count,
              data[scanOffset] == 0x0D,
              data[scanOffset + 1] == 0x0A {
            scanOffset += 2
        }
        guard scanOffset < data.count else {
            return .needsMoreData
        }
        let effectiveData = data[scanOffset...]

        let separator = Data("\r\n\r\n".utf8)
        guard let separatorRange = effectiveData.range(of: separator) else {
            return .needsMoreData
        }

        let headerData = effectiveData[effectiveData.startIndex..<separatorRange.lowerBound]
        // Validate UTF-8 encoding by round-tripping through String and back.
        // This rejects overlong encodings, lone surrogates, and other malformed
        // sequences that a lenient decoder might silently accept.
        guard let headerString = String(data: headerData, encoding: .utf8),
              Data(headerString.utf8) == Data(headerData) else {
            return .invalid(.invalidEncoding)
        }

        let lines = headerString.components(separatedBy: "\r\n")
        guard let requestLine = lines.first, !requestLine.isEmpty else {
            return .invalid(.emptyHeaderSection)
        }

        let parts = requestLine.split(separator: " ", maxSplits: 2, omittingEmptySubsequences: false)
        guard parts.count >= 2 else {
            return .invalid(.malformedRequestLine)
        }

        let method = String(parts[0])
        let target = String(parts[1])

        // RFC 9110 §9.1: method = token (tchar characters only).
        guard method.allSatisfy({ isTokenChar($0) }) else {
            return .invalid(.malformedRequestLine)
        }

        // RFC 9112 §2.3: HTTP-version = "HTTP/" DIGIT "." DIGIT
        guard parts.count >= 3 else {
            return .invalid(.invalidHTTPVersion)
        }
        let httpVersion = String(parts[2])
        guard isValidHTTPVersion(httpVersion) else {
            return .invalid(.invalidHTTPVersion)
        }

        // RFC 9112 §3.2: Validate request-target form.
        // origin-form: starts with "/"
        // absolute-form: starts with a scheme (e.g. "http://", "https://")
        // asterisk-form: "*" (only valid for OPTIONS)
        // authority-form: only valid for CONNECT
        if method == "CONNECT" {
            // authority-form: host:port — must contain a colon and not start with "/"
            if target.hasPrefix("/") || !target.contains(":") {
                return .invalid(.malformedRequestLine)
            }
        } else if method == "OPTIONS" && target == "*" {
            // asterisk-form is valid for OPTIONS
        } else if target.hasPrefix("/") {
            // origin-form — valid for all methods
        } else if target.lowercased().hasPrefix("http://") || target.lowercased().hasPrefix("https://") {
            // absolute-form — valid for all methods
        } else {
            return .invalid(.malformedRequestLine)
        }

        var headers: [String: String] = [:]
        var keyIndex: [String: String] = [:]  // lowercased -> original casing
        var contentLengthValue: Int64?
        var hostHeaderCount = 0
        var headerCount = 0
        for line in lines.dropFirst() where !line.isEmpty {
            headerCount += 1
            if headerCount > 100 {
                return .invalid(.tooManyHeaders)
            }
            // RFC 7230 §3.2.4: Reject obs-fold (continuation line starting with SP or HTAB)
            if line.first == " " || line.first == "\t" {
                return .invalid(.obsFoldDetected)
            }

            guard let colonIndex = line.firstIndex(of: ":") else {
                // RFC 9112 §5: A line with content but no colon is not a valid field line.
                return .invalid(.invalidFieldName)
            }

            let rawKey = line[line.startIndex..<colonIndex]

            // RFC 7230 §3.2.4: No whitespace is allowed between the field-name and colon.
            // Check this before the general token validation so we return the more
            // specific error (.whitespaceBeforeColon) instead of .invalidFieldName.
            if rawKey.contains(where: { $0 == " " || $0 == "\t" }) {
                return .invalid(.whitespaceBeforeColon)
            }

            // RFC 9110 §5.1: field-name = token (must be non-empty and contain only tchar)
            guard !rawKey.isEmpty, rawKey.allSatisfy({ isTokenChar($0) }) else {
                return .invalid(.invalidFieldName)
            }

            let key = String(rawKey)
            let lowerKey = key.lowercased()
            let value = String(line[line.index(after: colonIndex)...]).trimmingCharacters(in: .whitespaces)

            // RFC 9110 §5.5: field-value may only contain VCHAR (0x21-0x7E),
            // SP (0x20), HTAB (0x09), and obs-text (0x80-0xFF).
            // Reject any byte in 0x00-0x08, 0x0A-0x1F, and 0x7F.
            for scalar in value.unicodeScalars {
                let v = scalar.value
                if v <= 0x08 || (v >= 0x0A && v <= 0x1F) || v == 0x7F {
                    return .invalid(.invalidFieldValue)
                }
            }

            // RFC 7230 §3.3.3: Reject requests with Transfer-Encoding since this
            // server does not support chunked decoding. Silently ignoring it would
            // cause body framing mismatches (request smuggling).
            if lowerKey == "transfer-encoding" {
                return .invalid(.unsupportedTransferEncoding)
            }

            // Content-Length: validate and normalize to a single integer value.
            if lowerKey == "content-length" {
                do {
                    contentLengthValue = try validateContentLength(value, existing: contentLengthValue)
                } catch {
                    return .invalid(error)
                }
                // Store the resolved integer as the canonical header value,
                // not the raw (possibly comma-separated) form.
                let resolved = String(contentLengthValue!)
                if let existingKey = keyIndex["content-length"] {
                    headers[existingKey] = resolved
                } else {
                    headers[key] = resolved
                    keyIndex["content-length"] = key
                }
                continue
            }

            // Track Host header occurrences for RFC 9110 §7.2 validation.
            if lowerKey == "host" {
                hostHeaderCount += 1
            }

            // RFC 9110 §5.3: Combine duplicate field lines with comma-separated values.
            if let existingKey = keyIndex[lowerKey] {
                headers[existingKey] = "\(headers[existingKey]!), \(value)"
            } else {
                headers[key] = value
                keyIndex[lowerKey] = key
            }
        }

        // RFC 9110 §7.2: Reject requests with multiple Host headers (any version)
        // or missing Host header (HTTP/1.1 only).
        if hostHeaderCount > 1 {
            return .invalid(.multipleHostHeaders)
        }
        if httpVersion == "HTTP/1.1" && hostHeaderCount == 0 {
            return .invalid(.missingHostHeader)
        }

        let contentLength = contentLengthValue ?? 0

        let bodyOffset = data.distance(from: data.startIndex, to: separatorRange.upperBound)

        return .parsed(ParsedHeaders(
            method: method,
            target: target,
            httpVersion: httpVersion,
            headers: headers,
            contentLength: contentLength,
            bodyOffset: bodyOffset
        ))
    }

    /// Validates a Content-Length header value per RFC 9110 §8.6 / RFC 7230 §3.3.3.
    ///
    /// A Content-Length value may be a single number or a comma-separated list of
    /// identical values (e.g. "5, 5"). Each element must be a non-negative decimal
    /// integer (ASCII digits only — no +, ., 0x, etc.).
    ///
    /// - Parameters:
    ///   - value: The raw header value string.
    ///   - existing: A previously parsed Content-Length value, if any.
    /// - Returns: The validated content length as an integer.
    /// - Throws: ``HTTPRequestParseError/invalidContentLength`` or ``HTTPRequestParseError/conflictingContentLength``.
    static func validateContentLength(_ value: String, existing: Int64?) throws(HTTPRequestParseError) -> Int64 {
        let parts = value.split(separator: ",", omittingEmptySubsequences: false).map {
            $0.trimmingCharacters(in: .whitespaces)
        }
        guard let first = parts.first,
              !first.isEmpty,
              first.allSatisfy({ $0.isASCII && $0.isNumber }),
              let cl = Int64(first),
              cl >= 0
        else {
            throw HTTPRequestParseError.invalidContentLength
        }
        // All parts in a comma-separated list must represent the same integer value
        for part in parts.dropFirst() {
            guard !part.isEmpty,
                  part.allSatisfy({ $0.isASCII && $0.isNumber }),
                  let partValue = Int64(part),
                  partValue == cl
            else {
                throw HTTPRequestParseError.conflictingContentLength
            }
        }
        if let existing = existing, existing != cl {
            throw HTTPRequestParseError.conflictingContentLength
        }
        return cl
    }

    /// Validates that a string matches the HTTP-version format: `HTTP/DIGIT.DIGIT`.
    private static func isValidHTTPVersion(_ version: String) -> Bool {
        let prefix = "HTTP/"
        guard version.hasPrefix(prefix) else { return false }
        let rest = version.dropFirst(prefix.count)
        let parts = rest.split(separator: ".", maxSplits: 1)
        guard parts.count == 2,
              parts[0].count == 1, parts[0].first?.isASCII == true, parts[0].first?.isNumber == true,
              parts[1].count == 1, parts[1].first?.isASCII == true, parts[1].first?.isNumber == true
        else { return false }
        return true
    }

    /// Returns whether a character is a valid HTTP token character (RFC 9110 §5.6.2).
    ///
    /// `tchar = "!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" / "." /
    ///           "^" / "_" / "`" / "|" / "~" / DIGIT / ALPHA`
    private static func isTokenChar(_ c: Character) -> Bool {
        guard let ascii = c.asciiValue else { return false }
        switch ascii {
        case UInt8(ascii: "A")...UInt8(ascii: "Z"),
             UInt8(ascii: "a")...UInt8(ascii: "z"),
             UInt8(ascii: "0")...UInt8(ascii: "9"):
            return true
        case UInt8(ascii: "!"), UInt8(ascii: "#"), UInt8(ascii: "$"), UInt8(ascii: "%"),
             UInt8(ascii: "&"), UInt8(ascii: "'"), UInt8(ascii: "*"), UInt8(ascii: "+"),
             UInt8(ascii: "-"), UInt8(ascii: "."), UInt8(ascii: "^"), UInt8(ascii: "_"),
             UInt8(ascii: "`"), UInt8(ascii: "|"), UInt8(ascii: "~"):
            return true
        default:
            return false
        }
    }
}
